ledgerwallet/manifest_toml.py

```diff
@@ -0,0 +1,99 @@
+        else:
+            return self.dic[device]["dataSize"]
+
+    def get_application_flags(self, device) -> int:
+        if "flags" not in self.dic[device]:
+            return 0
+        else:
+            return int(self.dic[device]["flags"], 16)
+
+    def get_api_level(self, device) -> int:
+        level = self.dic[device]["apiLevel"]
+        if isinstance(level, int):
+            return int(level)
+        return int(level, 10)
+
+    def get_binary(self, device) -> str:
+        return os.path.join(self.path, self.dic[device]["binary"])
+
+    def has_api_level(self, device) -> bool:
+        if device == "0x33100004":
+            return self.dic[device].get("apiLevel") is not None
+        return False
+
+    def serialize_parameters(self, device) -> bytes:
+        parameters = []
+        for entry, value in self.dic.items():
+            if entry == "name":
+                parameters.append({"type_": "BOLOS_TAG_APPNAME", "value": value})
+            elif entry == "version":
+                parameters.append({"type_": "BOLOS_TAG_APPVERSION", "value": value})
+            elif entry == device:
+                for device_entry, device_value in self.dic[entry].items():
+                    if device_entry == "icon":
+                        parameters.append(
+                            {
+                                "type_": "BOLOS_TAG_ICON",
+                                "value": icon_from_file(device_value),
+                            }
+                        )
+                    elif device_entry == "derivationPath":
+                        derivation_paths: Dict[str, Optional[int]] = {
+                            "paths": None,
+                            "curve": None,
+                        }
+                        for derivation_entry in device_value:
+                            if derivation_entry == "curves":
+                                curves = 0
+                                for curve in device_value["curves"]:
+                                    if curve == "secp256k1":
+                                        curves |= params.CURVE_SECP256K1
+                                    elif curve == "prime256r1":
+                                        curves |= params.CURVE_PRIME256R1
+                                    elif curve == "ed25519":
+                                        curves |= params.CURVE_ED25519
+                                    elif curve == "bls12381g1":
+                                        curves |= params.CURVE_BLS12381G1
+                                    derivation_paths["curve"] = curves
+                            elif derivation_entry == "paths":
+                                derivation_paths["paths"] = device_value["paths"]
+                        parameters.append(
+                            {"type_": "BOLOS_TAG_DERIVEPATH", "value": derivation_paths}
+                        )
+        return params.AppParams.build(parameters)
```

Comment:
This code sounds very similar to `AppManifestJson.serialize_parameters` method, and should be factorized as much as possible (through private methods into `AppManifest`?).